sql/core/src/main/scala/org/apache/spark/sql/jdbc/PostgresDialect.scala

```diff
@@ -164,4 +172,78 @@ private object PostgresDialect extends JdbcDialect {
     s"TABLESAMPLE BERNOULLI" +
       s" (${(sample.upperBound - sample.lowerBound) * 100}) REPEATABLE (${sample.seed})"
   }
+
+  // CREATE INDEX syntax
+  // https://www.postgresql.org/docs/14/sql-createindex.html
+  override def createIndex(
+      indexName: String,
+      tableName: String,
+      columns: Array[NamedReference],
+      columnsProperties: util.Map[NamedReference, util.Map[String, String]],
+      properties: util.Map[String, String]): String = {
+    val columnList = columns.map(col => quoteIdentifier(col.fieldNames.head))
+    var indexPropertiesStr: String = ""
+    var hasIndexProperties: Boolean = false
+    var indexType = ""
+
+    if (!properties.isEmpty) {
+      var indexPropertyList: Array[String] = Array.empty
+      properties.asScala.foreach { case (k, v) =>
+        if (k.equals(SupportsIndex.PROP_TYPE)) {
+          if (v.equalsIgnoreCase("BTREE") || v.equalsIgnoreCase("HASH")) {
+            indexType = s"USING $v"
+          } else {
+            throw new UnsupportedOperationException(s"Index Type $v is not supported." +
+              " The supported Index Types are: BTREE and HASH")
+          }
+        } else {
+          hasIndexProperties = true
+          indexPropertyList = indexPropertyList :+ s"$k = $v"
+        }
+      }
+      if (hasIndexProperties) {
+        indexPropertiesStr += "WITH (" + indexPropertyList.mkString(", ") + ")"
+      }
+    }
+
+    s"CREATE INDEX ${quoteIdentifier(indexName)} ON ${quoteIdentifier(tableName)}" +
+      s" $indexType (${columnList.mkString(", ")}) $indexPropertiesStr"
+  }
+
+  // SHOW INDEX syntax
+  // https://www.postgresql.org/docs/14/view-pg-indexes.html
+  override def indexExists(
+      conn: Connection,
+      indexName: String,
+      tableName: String,
+      options: JDBCOptions): Boolean = {
+    val sql = s"SELECT * FROM pg_indexes WHERE tablename = '$tableName'"
+    try {
+      JdbcUtils.checkIfIndexExists(conn, indexName, sql, "indexname", options)
```

Comment:
I can not use `JdbcUtils.executeQuery(conn, options, sql)`. When I use it to execute query show index in Postgres, `rs.next()` throw exception `"The ResultSet is closed"`. That function return `resultset`, but it is closed when `statement` is closed in finally block. As in java 8 docs: [ResultSet](https://docs.oracle.com/javase/8/docs/api/java/sql/ResultSet.html#:~:text=A%20ResultSet%20object%20is%20automatically%20closed%20when%20the%20Statement%20object%20that%20generated%20it%20is%20closed%2C%20re-executed%2C%20or%20used%20to%20retrieve%20the%20next%20result%20from%20a%20sequence%20of%20multiple%20results.).
So I create `JdbcUtils.checkIfIndexExists` to execute the resultset in `JdbcUtils` for checking if the index exists, and just return boolean. WDYT, @huaxingao?

Reply:
How about moving the try catch to `JdbcUtils` too? So here we have
```
    val sql = s"SELECT * FROM pg_indexes WHERE tablename = ${quoteIdentifier(tableName)} AND" +
      s" indexname = '$indexName'"
    JdbcUtils.checkIfIndexExists(conn, sql, options)
```
In `JDBCUtils`, we have
```
  def checkIfIndexExists(
      conn: Connection,
      sql: String,
      options: JDBCOptions): Boolean = {
    val statement = conn.createStatement
    try {
      statement.setQueryTimeout(options.queryTimeout)
      val rs = statement.executeQuery(sql)
      rs.next
    } catch {
      case _: Exception =>
        logWarning("Cannot retrieved index info.")
        false
    } finally {
      statement.close()
    }
  }
```